src/EditorFeatures/Core/DocumentationComments/DocumentationCommentSuggestion.cs

```diff
@@ -65,34 +66,122 @@ public override Task OnProposalUpdatedAsync(SuggestionSessionBase session, Propo
         {
             try
             {
-                await providerInstance.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
-                await session.DisplayProposalAsync(Proposal, cancellationToken).ConfigureAwait(false);
-                return true;
+                await RunWithEnqueueActionAsync<bool>(
+                    "DisplayProposal",
+                    async () =>
+                    {
+                        await _suggestionSession!.DisplayProposalAsync(proposal, cancellationToken).ConfigureAwait(false);
+                        return true;
+                    },
+                    cancellationToken).ConfigureAwait(false);
+
+                Logger.Log(FunctionId.Copilot_Generate_Documentation_Displayed, logLevel: LogLevel.Information);
             }
             catch (OperationCanceledException)
             {
                 Logger.Log(FunctionId.Copilot_Generate_Documentation_Canceled, logLevel: LogLevel.Information);
             }
+        }
+
+        /// <summary>
+        /// Dismisses the session if the proposal we generated was invalid.
+        /// Needs to dispose of the IntelliCodeCompletionsDisposable so we no longer have exclusive right to
+        /// display any grey text.
+        /// </summary>
+        private async Task DismissSuggestionSessionAsync(CancellationToken cancellationToken)
+        {
+            await DisposeIntelliCodeCompletionsDisposableAsync().ConfigureAwait(false);
+            await RunWithEnqueueActionAsync<bool>(
+                "DismissSuggestionSession",
+                async () =>
+                {
+                    await _suggestionSession!.DismissAsync(ReasonForDismiss.DismissedDueToInvalidProposal, cancellationToken).ConfigureAwait(false);
+                    return true;
+                },
+                cancellationToken).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// In general, calls to a SuggestionManager or SuggestionSession need to be wrapped in an EnqueueAction.
+        /// This is the pattern recommended by VS Platform to avoid races.
+        /// Pattern from platform shown here:
+        /// https://devdiv.visualstudio.com/DevDiv/_git/IntelliCode-VS?path=/src/VSIX/IntelliCode.VSIX/SuggestionService/AmbientAI/SuggestionProviderForAmbientAI.cs
+        /// </summary>
+        private async Task<T> RunWithEnqueueActionAsync<T>(string description, Func<Task<T>> action, CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(SuggestionManager);
+
+            var taskCompletionSource = new TaskCompletionSource<T>();
+
+            await providerInstance.ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            SuggestionManager.EnqueueAction(description, async () =>
+            {
+                try
+                {
+                    var result = await action().ConfigureAwait(false);
```

Comment:
Since all you do here is mark a TCS complete, you can potentially improve efficiency by using [ConfigureAwaitRunInline](https://learn.microsoft.com/en-us/dotnet/api/microsoft.visualstudio.threading.awaitextensions.configureawaitruninline?view=visualstudiosdk-2022) here instead.
Just be sure to combine it with my other suggestion to prevent _other_ continuations from being inlined.